src/irmin-pack/irmin_pack.ml

```diff
@@ -200,10 +205,19 @@ module Atomic_write (K : Irmin.Type.S) (V : Irmin.Hash.S) = struct
         (aux [@tailcall]) (off ++ Int64.(of_int @@ (len + V.hash_size))) k
     in
     (aux [@tailcall]) 0L @@ fun () ->
-    { cache; index; block; w = watches; lock = Lwt_mutex.create () }
+    {
+      cache;
+      index;
+      block;
+      w = watches;
+      lock = Lwt_mutex.create ();
+      counter = 1;
+    }
 
   let (`Staged unsafe_v) =
-    with_cache ~clear:unsafe_clear ~v:(fun () -> unsafe_v) "store.branches"
+    with_cache ~clear:unsafe_clear ~valid ~incr_counter
```

Comment:
can you try to merge `valid` and `incr_counter` to simplify that API a bit? For instance, you could define `valid` as:

```ocaml
let valid' t =   if valid t then (incr_counter t; true) else false
```

or something similar